clientconn_test.go

```diff
@@ -63,72 +63,68 @@ func assertState(wantState connectivity.State, cc *ClientConn) (connectivity.Sta
 
 func TestDialWithMultipleBackendsNotSendingServerPreface(t *testing.T) {
 	defer leakcheck.Check(t)
-	numServers := 2
-	servers := make([]net.Listener, numServers)
-	var err error
-	for i := 0; i < numServers; i++ {
-		servers[i], err = net.Listen("tcp", "localhost:0")
+	server1, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening. Err: %v", err)
+	}
+	defer server1.Close()
+	server1Addr := resolver.Address{Addr: server1.Addr().String()}
+	server1Done := make(chan struct{})
+	// 1st server accepts the connection and immediately closes it.
+	go func() {
+		defer close(server1Done)
+		conn, err := server1.Accept()
 		if err != nil {
-			t.Fatalf("Error while listening. Err: %v", err)
+			t.Errorf("Error while accepting. Err: %v", err)
+			return
 		}
+		defer conn.Close()
+	}()
+
+	server2, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening. Err: %v", err)
 	}
-	dones := make([]chan struct{}, numServers)
-	for i := 0; i < numServers; i++ {
-		dones[i] = make(chan struct{})
-	}
-	for i := 0; i < numServers; i++ {
-		go func(i int) {
-			defer func() {
-				close(dones[i])
-			}()
-			conn, err := servers[i].Accept()
-			if err != nil {
-				t.Errorf("Error while accepting. Err: %v", err)
-				return
-			}
-			defer conn.Close()
-			switch i {
-			case 0: // 1st server accepts the connection and immediately closes it.
-			case 1: // 2nd server accepts the connection and sends settings frames.
-				framer := http2.NewFramer(conn, conn)
-				if err := framer.WriteSettings(http2.Setting{}); err != nil {
-					t.Errorf("Error while writing settings frame. %v", err)
-					return
-				}
-				conn.SetDeadline(time.Now().Add(time.Second))
-				buf := make([]byte, 1024)
-				for { // Make sure the connection stays healthy.
-					_, err = conn.Read(buf)
-					if err == nil {
-						continue
-					}
-					if nerr, ok := err.(net.Error); !ok || !nerr.Timeout() {
-						t.Errorf("Server expected the conn.Read(_) to timeout instead got error: %v", err)
-					}
-					return
-				}
-			}
-		}(i)
-	}
+	defer server2.Close()
+	server2Done := make(chan struct{})
+	server2Addr := resolver.Address{Addr: server2.Addr().String()}
+	// 2nd server accepts the connection and sends settings frames.
+	go func() {
+		defer close(server2Done)
+		conn, err := server2.Accept()
+		if err != nil {
+			t.Errorf("Error while accepting. Err: %v", err)
+			return
+		}
+		defer conn.Close()
+		framer := http2.NewFramer(conn, conn)
+		if err := framer.WriteSettings(http2.Setting{}); err != nil {
+			t.Errorf("Error while writing settings frame. %v", err)
+			return
+		}
+		conn.SetDeadline(time.Now().Add(time.Second))
+		go keepReading(conn)
```

Comment:
I think in the original idea for the test, this should not be done in a goroutine.  Otherwise you just immediately close the conn from the defer above, and the connection will probably die before the client even considers it READY.

If the real goal of the test is to make sure the client dials the second address in the list after the first one failed pre-preface, then I don't see any reason we need to mess with the framer.